server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/routes/MailRepositoriesRoutes.java

```diff
@@ -169,33 +176,72 @@ public void defineGetMailRepositories() {
     }
 
+    @GET
+    @Produces("application/json, message/rfc822")
     @Path("/{encodedUrl}/mails/{mailKey}")
-    @ApiOperation(value = "Retrieving a specific mail details")
+    @ApiOperation(value = "Retrieving a specific mail details (this endpoint can accept both \"application/json\" or \"message/rfc822\")")
     @ApiResponses(value = {
         @ApiResponse(code = HttpStatus.OK_200, message = "The list of all mails in a repository", response = List.class),
         @ApiResponse(code = HttpStatus.INTERNAL_SERVER_ERROR_500, message = "Internal server error - Something went bad on the server side."),
         @ApiResponse(code = HttpStatus.NOT_FOUND_404, message = "Not found - Could not retrieve the given mail.")
     })
     public void defineGetMail() {
-        service.get(MAIL_REPOSITORIES + "/:encodedUrl/mails/:mailKey", (request, response) -> {
-            String url = decodedRepositoryUrl(request);
-            String mailKey = request.params("mailKey");
-            try {
-                return repositoryStoreService.retrieveMail(url, mailKey)
-                    .orElseThrow(() -> ErrorResponder.builder()
-                        .statusCode(HttpStatus.NOT_FOUND_404)
-                        .type(ErrorResponder.ErrorType.NOT_FOUND)
-                        .message("Could not retrieve " + mailKey)
-                        .haltError());
-            } catch (MailRepositoryStore.MailRepositoryStoreException | MessagingException e) {
-                throw ErrorResponder.builder()
-                    .statusCode(HttpStatus.INTERNAL_SERVER_ERROR_500)
-                    .type(ErrorResponder.ErrorType.SERVER_ERROR)
-                    .cause(e)
-                    .message("Error while retrieving mail")
-                    .haltError();
-            }
-        }, jsonTransformer);
+
+        service.get(MAIL_REPOSITORIES + "/:encodedUrl/mails/:mailKey", Constants.JSON_CONTENT_TYPE,
+            (request, response) -> getMailAsJson(decodedRepositoryUrl(request), request.params("mailKey")), jsonTransformer);
+
+        service.get(MAIL_REPOSITORIES + "/:encodedUrl/mails/:mailKey", Constants.RFC822_CONTENT_TYPE,
+            (request, response) -> writeMimeMessage(
+                getMailAsMimeMessage(decodedRepositoryUrl(request), request.params("mailKey")),
+                response.raw()));
+    }
+
+    private Object writeMimeMessage(MimeMessage mimeMessage, HttpServletResponse rawResponse) throws MessagingException, IOException {
+        rawResponse.setContentType(Constants.RFC822_CONTENT_TYPE);
+        rawResponse.setHeader("Content-Length", String.valueOf(computeExactSize(mimeMessage)));
+        mimeMessage.writeTo(rawResponse.getOutputStream());
+        return rawResponse;
+    }
+
+    private long computeExactSize(MimeMessage mimeMessage) throws IOException, MessagingException {
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        mimeMessage.writeTo(byteArrayOutputStream);
```

Comment:
do we really need to load the message in memory to compute its size?

Reply:
Yes: we should know the size prior to the download to start and the getSize method is not accurate (I had test failures)

Reply:
ok that's not what I meant, but let's stick with that